### PR TITLE
remove GetLayerByID from ImageService interface

### PR DESCRIPTION
**Pull request by thaJeztah**
upstreams:

- https://github.com/rumpl/moby/pull/45
- https://github.com/rumpl/moby/pull/63

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
sigh.. hcsshim errors are so informative;

```
=== FAIL: github.com/docker/docker/integration/container TestLogs/driver_json-file/without_tty/only_stdout (1.49s)
    logs_test.go:132: assertion failed: error is not nil: Error response from daemon: failed to create task for container: failed to create shim task: hcs::CreateComputeSystem de93db4b975156762f042f0e542cafcbba5cd19eac7b964dc979aa8e427bdc0f: The parameter is incorrect.: unknown
```

**Review by vvoland — APPROVED** [review]

**Review by rumpl — APPROVED** [review]

File: daemon/daemon.go [hunk 1]
```diff
@@ -259,12 +264,14 @@ func (daemon *Daemon) restore() error {
...
-			if err != nil {
-				log.WithError(err).Error("failed to load container mount")
-				return
+			if accessor, ok := daemon.imageService.(layerAccessor); ok {
+				rwlayer, err := accessor.GetLayerByID(c.ID)
+				if err != nil {
+					log.WithError(err).Error("failed to load container mount")
+					return
+				}
+				c.RWLayer = rwlayer
```
Comment (thaJeztah) [inline daemon/daemon.go:273]: Discussing this with @rumpl - it's kinda ugly that we're mutating the container that we received from `c, err := daemon.load(id)` here (although this was already the case before this PR)

Wondering if `c, err := daemon.load(id)` should just return the container with the right options already set (or if needed some wrapper).

Comment (rumpl) [reply daemon/daemon.go:273]: I wonder why don't we use the layer store directly? `GetLayerByID` is literally a one-line wrapper around the layer store 

https://github.com/moby/moby/blob/ad9d70b0e6833b746eecccc08a17459ba8bb31c4/daemon/images/service.go#L140-L142

Comment (thaJeztah) [reply daemon/daemon.go:273]: I updated one instance, but (saw you also tried) looks like we're using it elsewhere. I also fixed two comments that were no longer correct, and added a `Warn()` log for an unhandled error when releasing the layer;

https://github.com/moby/moby/compare/14cc4c994e7f653512f95a7b31fc90cc0f09a0ba..1f8b444f1bdd94d8a19116753cf83607d7479029


One thing we should probably look into is that in some cases we call `i.layerstore.ReleaseRWLayer()` directly, and in other cases we call `i.ReleaseLayer()`

**Review by vvoland — APPROVED** [review]
I think we should log the second error instead 😄 
LGTM once addressed

File: daemon/images/image_exporter.go [hunk 2]
```diff
@@ -17,6 +19,28 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
...
+func (i *ImageService) PerformWithBaseFS(ctx context.Context, c *container.Container, fn func(root string) error) error {
+	rwlayer, err := i.layerStore.GetRWLayer(c.ID)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			err2 := i.ReleaseLayer(rwlayer)
+			if err2 != nil {
+				logrus.WithError(err).WithField("container", c.ID).Warn("Failed to release layer")
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply daemon/images/image_exporter.go]: whoops, yes, thanks!

**Review by vvoland — APPROVED** [review]

**Pull request merged**